File: alchemist/alchemist-swingui/src/main/java/it/unibo/alchemist/boundary/monitor/generic/AbstractFXDisplay.java
package it.unibo.alchemist.boundary.monitor.generic;

import it.unibo.alchemist.boundary.gui.effects.EffectGroup;
import it.unibo.alchemist.boundary.gui.utility.DataFormatFactory;
import it.unibo.alchemist.boundary.interfaces.FXOutputMonitor;
import it.unibo.alchemist.boundary.wormhole.implementation.Wormhole2D;
import it.unibo.alchemist.boundary.wormhole.interfaces.BidimensionalWormhole;
import it.unibo.alchemist.model.implementations.times.DoubleTime;
import it.unibo.alchemist.model.interfaces.Concentration;
import it.unibo.alchemist.model.interfaces.Environment;
import it.unibo.alchemist.model.interfaces.Node;
import it.unibo.alchemist.model.interfaces.Position;
import it.unibo.alchemist.model.interfaces.Position2D;
import it.unibo.alchemist.model.interfaces.Reaction;
import it.unibo.alchemist.model.interfaces.Time;
import java.awt.*;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.lang.reflect.Modifier;
import java.util.Collection;
import java.util.Optional;
import java.util.concurrent.ConcurrentLinkedQueue;
import java.util.concurrent.Semaphore;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.stream.Collectors;
import java.util.stream.Stream;
import javafx.application.Platform;
import javafx.collections.FXCollections;
import javafx.collections.ObservableList;
import javafx.scene.canvas.Canvas;
import javafx.scene.canvas.GraphicsContext;
import javafx.scene.input.ClipboardContent;
import javafx.scene.input.DataFormat;
import javafx.scene.input.DragEvent;
import javafx.scene.input.Dragboard;
import javafx.scene.input.InputEvent;
import javafx.scene.input.MouseEvent;
import javafx.scene.input.TransferMode;
import org.jetbrains.annotations.Nullable;

/**
 * Base abstract class for each display able to graphically represent a 2D space and simulation.
 *
 * @param <T> The type which describes the {@link Concentration} of a molecule
 * @param <P> The type of position
 */
public abstract class AbstractFXDisplay<T, P extends Position2D<? extends P>> extends Canvas implements FXOutputMonitor<T, P> {
    /**
     * The default frame rate.
     */
    public static final byte DEFAULT_FRAME_RATE = 60;
    /**
     * The default time per frame.
     */
    public static final double TIME_STEP = 1 / DEFAULT_FRAME_RATE;
    /**
     * Default number of steps.
     */
    protected static final int DEFAULT_NUMBER_OF_STEPS = 1;
    /**
     * Position {@code DataFormat}.
     */
    protected static final DataFormat POSITION_DATA_FORMAT = DataFormatFactory.getDataFormat(Position.class);
    /**
     * Default serial version UID.
     */
    private static final long serialVersionUID = 1L;
    /**
     * The default view status.
     */
    private static final FXOutputMonitor.ViewStatus DEFAULT_VIEW_STATUS = FXOutputMonitor.ViewStatus.PAN;
    private static final String GET_X_METHOD_NAME = "getX";
    private static final String GET_Y_METHOD_NAME = "getY";
    private final ObservableList<EffectGroup> effectStack;
    private final Semaphore mutex = new Semaphore(1);
    private final AtomicBoolean mayRender = new AtomicBoolean(true);
    private int step;
    private BidimensionalWormhole<P> wormhole;
    private volatile boolean firstTime;
    private boolean realTime;
    private volatile ConcurrentLinkedQueue<Runnable> commandQueue;
    private FXOutputMonitor.ViewStatus viewStatus;

    /**
     * Default constructor. The number of steps is set to default ({@value #DEFAULT_NUMBER_OF_STEPS}).
     */
    public AbstractFXDisplay() {
        this(DEFAULT_NUMBER_OF_STEPS);
    }

    /**
     * Main constructor. It lets the developer specify the number of steps.
     *
     * @param steps the number of steps
     * @see #setStep(int)
     */
    public AbstractFXDisplay(final int steps) {
        super();
        this.firstTime = true;
        this.realTime = false;
        this.effectStack = FXCollections.observableArrayList();
        setStep(steps);
        this.commandQueue = new ConcurrentLinkedQueue<>();
        enableEventReceiving();
        setStyle("-fx-background-color: #FFF;");
        initMouseListener(); // NOPMD - the method is meant to be overridable
        setViewStatus(DEFAULT_VIEW_STATUS); // NOPMD - the method is meant to be overridable
    }

    /**
     * Enables {@link MouseEvent} receiving by enabling Focus and requesting it.
     */
    private void enableEventReceiving() {
        setFocusTraversable(true);
        setFocused(true);
    }

    /**
     * Initializes the mouse interaction to the {@link Canvas}.
     * <p>
     * Should be overridden to implement mouse interaction with the GUI.
     */
    protected void initMouseListener() {
        // TODO
        setOnMouseClicked(event -> {
            switch (event.getButton()) {
                case PRIMARY:
                    // TODO Handle primary button
                    break;
                case SECONDARY:
                    // TODO Handle secondary button
                    break;
                default:
                    // Do nothing
                    break;
            }
        });

        setOnDragDetected(event -> {
            switch (getViewStatus()) {
                case PAN:
                    startEnvironmentDragNDrop(event);
                    break;
                case MOVING:
                    startNodeDragNDrop(event);
                    break;
                // TODO
                default:
                    break;
            }
        });

        setOnDragEntered(event -> {
            switch (getViewStatus()) {
                case PAN:
                    onEnvironmentDragEntered(event);
                    break;
                case MOVING:
                    onNodeDragEntered(event);
                    break;
                // TODO
                default:
                    break;
            }
        });
    }

    @Override
    public final ViewStatus getViewStatus() {
        return this.viewStatus;
    }

    @Override
    public final void setViewStatus(final ViewStatus viewStatus) {
        this.viewStatus = viewStatus;
    }

    /**
     * The method returns the {@link Position} in the {@link Environment} of the given {@code Event}, if any.
     *
     * @param event the event to check
     * @return the position, if any
     */
    protected final Optional<P> getEventPosition(final InputEvent event) {
        final BidimensionalWormhole<P> wormhole = getWormhole();
        final Method[] methods = event.getClass().getMethods();
        Optional<Method> getX = Optional.empty();
        Optional<Method> getY = Optional.empty();

        for (final Method method : methods) {
            final int modifier = method.getModifiers();
            if (Modifier.isPublic(modifier) && !Modifier.isAbstract(modifier)) {
                final String name = method.getName();
                if (name.equals(GET_X_METHOD_NAME)) {
                    getX = Optional.of(method);
                } else if (name.equals(GET_Y_METHOD_NAME)) {
                    getY = Optional.of(method);
                }
                if (getX.isPresent() && getY.isPresent()) {
                    break;
                }
            }
        }

        if (wormhole != null && getX.isPresent() && getY.isPresent()) {
            try {
                final Number x = (Number) getX.get().invoke(event);
                final Number y = (Number) getY.get().invoke(event);
                return Optional.of(wormhole.getEnvPoint(new Point(x.intValue(), y.intValue())));
            } catch (final IllegalAccessException | InvocationTargetException e) {
                return Optional.empty();
            }
        } else {
            return Optional.empty();
        }
    }

    /**
     * The method is meant to handle what should the monitor do when a drag'n'drop of a {@link Node} starts.
     *
     * @param event the event
     */
    protected void startNodeDragNDrop(final MouseEvent event) {
        // TODO
    }

    /**
     * The method is meant to handle what should the monitor do when a pan gesture in the {@link Environment} starts.
     *
     * @param event the event
     */
    protected void startEnvironmentDragNDrop(final MouseEvent event) {
        final Optional<P> position = getEventPosition(event);
        position.ifPresent(p -> {
            final Dragboard dragboard = startDragAndDrop(TransferMode.MOVE);
            final ClipboardContent content = new ClipboardContent();
            content.put(POSITION_DATA_FORMAT, p);
            dragboard.setContent(content);
        });

        event.consume();
    }

    /**
     * The method is meant to handle what should the monitor do when the mouse enters in a valid position during a pan gesture in the {@link Environment}.
     *
     * @param event the event
     */
    protected void onEnvironmentDragEntered(final DragEvent event) {
//        final BidimensionalWormhole wormhole = getWormhole();
//        if (wormhole != null) {
//            final Position previousMousePos = (Position) event.getDragboard().getContent(POSITION_DATA_FORMAT);
//            repaint();
//        }
        // TODO
    }

    /**
     * The method is meant to handle what should the monitor do when the mouse enters in a valid position during a drag'n'drop gesture of a {@link Node}.
     *
     * @param event the event
     */
    protected void onNodeDragEntered(final DragEvent event) {
        // TODO
    }

    @Override
    public int getStep() {
        return this.step;
    }

    /**
     * {@inheritDoc}
     *
     * @throws IllegalArgumentException if the step is not bigger than 0
     */
    @Override
    public void setStep(final int step) throws IllegalArgumentException {
        if (step <= 0) {
            throw new IllegalArgumentException("The parameter must be a positive integer");
        }
        this.step = step;
    }

    @Override
    public boolean isRealTime() {
        return this.realTime;
    }

    @Override
    public void setRealTime(final boolean realTime) {
        this.realTime = realTime;
    }

    @Override
    public final void repaint() {
        mutex.acquireUninterruptibly();
        if (mayRender.get() && getWormhole() != null && isVisible() && !isDisabled()) {
            mayRender.set(false);
            Platform.runLater(() -> {
                commandQueue.forEach(Runnable::run);
                mayRender.set(true);
            });
        }
        mutex.release();
    }

    /**
     * Changes the background of the specified {@code GraphicsContext}.
     *
     * @param graphicsContext the graphic component to draw on
     * @param environment     the {@code Environment} that contains the data to pass to {@code Effects}
     * @return a function of what to do to draw the background
     * @see #repaint()
     */
    protected Runnable drawBackground(final GraphicsContext graphicsContext, final Environment<T, P> environment) {
        return () -> graphicsContext.clearRect(0, 0, getWidth(), getHeight());
    }

    /**
     * Getter method for the wormhole the converts {@link Environment} positions to GUI positions.
     *
     * @return the wormhole object
     */
    @Nullable
    protected BidimensionalWormhole<P> getWormhole() {
        return this.wormhole;
    }

    @Override
    public void addEffects(final Collection<EffectGroup> effects) {
        this.effectStack.addAll(effects);
    }

    @Override
    public void addEffectGroup(final EffectGroup effects) {
        this.effectStack.add(effects);
    }

    @Override
    public Collection<EffectGroup> getEffects() {
        return this.effectStack;
    }

    @Override
    public void setEffects(final Collection<EffectGroup> effects) {
        this.effectStack.clear();
        this.effectStack.addAll(effects);
    }

    @Override
    public void initialized(final Environment<T, P> environment) {
        stepDone(environment, null, new DoubleTime(), 0);
    }

    @Override
    public void stepDone(final Environment<T, P> environment, final Reaction<T> reaction, final Time time, final long step) {
        if (firstTime) {
            synchronized (this) {
                if (firstTime) {
                    init(environment);
                    update(environment, time);
                }
            }
        } else {
            update(environment, time);
        }
    }

    /**
     * The method initializes everything is not initializable before first step.
     *
     * @param environment the {@code Environment}
     */
    protected void init(final Environment<T, P> environment) {
        wormhole = new Wormhole2D<>(environment, this);
        wormhole.center();
        wormhole.optimalZoom();
        firstTime = false;
        System.currentTimeMillis();
    }

    @Override
    public void finished(final Environment<T, P> environment, final Time time, final long step) {
        update(environment, time);
        firstTime = true;
    }

    /**
     * Updates parameter for correct {@code Environment} representation.
     *
     * @param environment the {@code Environment}
     * @param time        the current {@code Time} of simulation
     */
    private void update(final Environment<T, P> environment, final Time time) {
        if (Thread.holdsLock(environment)) {
            time.toDouble();
            final GraphicsContext graphicsContext = this.getGraphicsContext2D();
            final Stream<Runnable> background = Stream.of(drawBackground(graphicsContext, environment));
            final Stream<Runnable> effects = getEffects()
                    .stream()
                    .map(group -> group.computeDrawCommands(environment))
                    .flatMap(Collection::stream)
                    .map(cmd -> () -> cmd.accept(graphicsContext, getWormhole()));
            commandQueue = Stream
                    .concat(background, effects)
                    .collect(Collectors.toCollection(ConcurrentLinkedQueue::new));
            repaint();
        } else {
            throw new IllegalStateException("Only the simulation thread can dictate GUI updates");
        }
    }
}

File: alchemist/alchemist-fxui/src/main/kotlin/it/unibo/alchemist/boundary/monitors/AbstractFXDisplay.kt
/*
 * Copyright (C) 2010-2018, Danilo Pianini and contributors listed in the main
 * project's alchemist/build.gradle file.
 *
 * This file is part of Alchemist, and is distributed under the terms of the
 * GNU General Public License, with a linking exception, as described in the file
 * LICENSE in the Alchemist distribution's top directory.
 */

package it.unibo.alchemist.boundary.monitors

import edu.umd.cs.findbugs.annotations.SuppressFBWarnings
import it.unibo.alchemist.boundary.clear
import it.unibo.alchemist.boundary.gui.effects.EffectGroup
import it.unibo.alchemist.boundary.gui.utility.DataFormatFactory
import it.unibo.alchemist.boundary.interactions.InteractionManager
import it.unibo.alchemist.boundary.interfaces.DrawCommand
import it.unibo.alchemist.boundary.interfaces.FXOutputMonitor
import it.unibo.alchemist.boundary.jfx.events.keyboard.KeyboardActionListener
import it.unibo.alchemist.boundary.wormhole.implementation.ExponentialZoomManager
import it.unibo.alchemist.boundary.wormhole.interfaces.Wormhole2D
import it.unibo.alchemist.boundary.wormhole.interfaces.ZoomManager
import it.unibo.alchemist.model.implementations.times.DoubleTime
import it.unibo.alchemist.model.interfaces.Concentration
import it.unibo.alchemist.model.interfaces.Environment
import it.unibo.alchemist.model.interfaces.Position
import it.unibo.alchemist.model.interfaces.Position2D
import it.unibo.alchemist.model.interfaces.Reaction
import it.unibo.alchemist.model.interfaces.Time
import it.unibo.alchemist.wormhole.implementation.WormholeFX
import javafx.application.Platform
import javafx.beans.value.ChangeListener
import javafx.collections.FXCollections
import javafx.collections.ObservableList
import javafx.scene.Group
import javafx.scene.Node
import javafx.scene.canvas.Canvas
import javafx.scene.layout.Pane
import java.util.Queue
import java.util.concurrent.ConcurrentLinkedQueue
import java.util.concurrent.Semaphore
import java.util.concurrent.atomic.AtomicBoolean
import java.util.stream.Collectors
import java.util.stream.Stream

/**
 * Base abstract class for each display able to graphically represent a 2D space and simulation.
 *
 * @param <T> The type which describes the [Concentration] of a molecule
 * @param <P> The type of position
 */
@SuppressFBWarnings(
    "NP_NONNULL_FIELD_NOT_INITIALIZED_IN_CONSTRUCTOR",
    "Field is initialized in the initialize function"
)
abstract class AbstractFXDisplay<T, P : Position2D<P>>
/**
 * Main constructor. It lets the developer specify the number of steps.
 *
 * @param steps the number of steps
 * @see .setStep
 */
@JvmOverloads constructor(steps: Int = DEFAULT_NUMBER_OF_STEPS) : Pane(), FXOutputMonitor<T, P> {

    private val effectStack: ObservableList<EffectGroup<P>> = FXCollections.observableArrayList()
    private val mutex = Semaphore(1)
    private val mayRender = AtomicBoolean(true)
    private var step: Int = 0
    @Volatile private var firstTime: Boolean = false
    private var realTime: Boolean = false
    @Volatile private var commandQueue: ConcurrentLinkedQueue<() -> Unit> = ConcurrentLinkedQueue()
    private var viewStatus = DEFAULT_VIEW_STATUS
    private lateinit var wormhole: Wormhole2D<P>
    private lateinit var zoomManager: ZoomManager
    private val interactions: InteractionManager<T, P> by lazy { InteractionManager(this) }
    private val effectsCanvas = Canvas()
    /**
     * Group dedicated for painting the background.
     */
    protected val background = Group()

    init {
        firstTime = true
        setStep(steps)
        val repaintOnResize = ChangeListener<Number> { _, _, _ ->
            repaint()
        }
        widthProperty().addListener(repaintOnResize)
        heightProperty().addListener(repaintOnResize)
        effectsCanvas.isMouseTransparent = true
        effectsCanvas.widthProperty().bind(widthProperty())
        effectsCanvas.heightProperty().bind(heightProperty())
        children.addAll(background, effectsCanvas, interactions.canvases)
    }

    override fun getViewStatus(): FXOutputMonitor.ViewStatus {
        return this.viewStatus
    }

    override fun setViewStatus(viewStatus: FXOutputMonitor.ViewStatus) {
        this.viewStatus = viewStatus
    }

    override fun getStep(): Int {
        return this.step
    }

    /**
     * {@inheritDoc}.
     *
     * @throws IllegalArgumentException if the step is not bigger than 0
     */
    @Throws(IllegalArgumentException::class)
    override fun setStep(step: Int) {
        if (step <= 0) {
            throw IllegalArgumentException("The parameter must be a positive integer")
        }
        this.step = step
    }

    override fun isRealTime(): Boolean {
        return this.realTime
    }

    override fun setRealTime(realTime: Boolean) {
        this.realTime = realTime
    }

    override fun repaint() {
        mutex.acquireUninterruptibly()
        if (mayRender.get() && isVisible && !isDisabled) {
            mayRender.set(false)
            Platform.runLater {
                commandQueue.forEach { it() }
                mayRender.set(true)
            }
        }
        interactions.repaint()
        mutex.release()
    }

    override fun addEffects(effects: Collection<EffectGroup<P>>) {
        this.effectStack.addAll(effects)
    }

    override fun addEffectGroup(effects: EffectGroup<P>) {
        this.effectStack.add(effects)
    }

    override fun getEffects(): Collection<EffectGroup<P>> {
        return this.effectStack
    }

    override fun setEffects(effects: Collection<EffectGroup<P>>) {
        this.effectStack.clear()
        this.effectStack.addAll(effects)
    }

    override fun getKeyboardListener(): KeyboardActionListener = interactions.keyboardListener

    override fun initialized(environment: Environment<T, P>) {
        init(environment)
        stepDone(environment, null, DoubleTime(), 0)
    }

    override fun stepDone(environment: Environment<T, P>, reaction: Reaction<T>?, time: Time, step: Long) {
        update(environment, time)
    }

    /**
     * The method initializes everything that is not initializable before first step.
     * Inheriting classes that override this function should always call super.init to ensure proper initialization.
     *
     * @param environment the `Environment`
     */
    protected open fun init(environment: Environment<T, P>) {
        wormhole = createWormhole(environment)
        zoomManager = createZoomManager(wormhole)
        interactions.setWormhole(wormhole)
        interactions.setZoomManager(zoomManager)
        wormhole.center()
        wormhole.optimalZoom()
        zoomManager.zoom = wormhole.zoom
        firstTime = false
        System.currentTimeMillis()
    }

    /**
     * Creates a wormhole for this monitor.
     * Subclasses that make use of their own wormholes can set them through this method.
     *
     * @param environment the current environment.
     * @returns the wormhole.
     */
    protected open fun createWormhole(environment: Environment<T, P>): WormholeFX<P> =
        WormholeFX(environment, this)

    /**
     * Creates a zoom manager for this monitor.
     * Subclasess that make use of their own zoom managers can set them through this method.
     *
     * @param wormhole the current wormhole.
     * @returns the zoom manager.
     */
    protected open fun createZoomManager(wormhole: Wormhole2D<P>): ZoomManager =
        ExponentialZoomManager(wormhole.zoom, ExponentialZoomManager.DEF_BASE)

    override fun finished(environment: Environment<T, P>, time: Time, step: Long) {
        update(environment, time)
        firstTime = true
    }

    /**
     * Updates parameter for correct `Environment` representation.
     *
     * @param environment the `Environment`
     * @param time the current `Time` of simulation
     */
    @SuppressFBWarnings("RV_RETURN_VALUE_IGNORED_NO_SIDE_EFFECT", "False positive")
    private fun update(environment: Environment<T, P>, time: Time) {
        if (Thread.holdsLock(environment)) {
            time.toDouble()
            interactions.environment = environment
//            TODO: Future optimization -- Let the simulation (or the environment, probably both)
//            expose the last moment at which a change in position occurred. This way, we don't
//            need to constantly regenerate the position map.
            interactions.nodes = environment.nodes.associateWith(environment::getPosition)
            val graphicsContext = effectsCanvas.graphicsContext2D
            val clearEffects = Stream.of(effectsCanvas::clear)
            val drawEffects = effects
                .stream()
                .map<Queue<DrawCommand<P>>> { group -> group.computeDrawCommands(environment) }
                .flatMap { it.stream() }
                .map { cmd -> { cmd.accept(graphicsContext, wormhole) } }
            commandQueue = Stream
                .concat(clearEffects, drawEffects)
                .collect(Collectors.toCollection { ConcurrentLinkedQueue<() -> Unit>() })
            interactions.simulationStep()
            repaint()
        } else {
            throw IllegalStateException("Only the simulation thread can dictate GUI updates")
        }
    }

    companion object {
        /**
         * The default frame rate.
         */
        const val DEFAULT_FRAME_RATE: Byte = 60
        /**
         * The default time per frame.
         */
        const val DEFAULT_TIME_STEP = (1 / DEFAULT_FRAME_RATE).toDouble()
        /**
         * Default number of steps.
         */
        const val DEFAULT_NUMBER_OF_STEPS = 1
        /**
         * Position `DataFormat`.
         */
        protected val POSITION_DATA_FORMAT = DataFormatFactory.getDataFormat(Position::class.java)
        /**
         * Default serial version UID.
         */
        private const val serialVersionUID = 1L
        /**
         * The default view status.
         */
        private val DEFAULT_VIEW_STATUS = FXOutputMonitor.ViewStatus.PANNING
    }

    /**
     * @inheritDoc.
     */
    override fun getNode() = this as Node
}
